staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,342 @@
+	}
+	streamer = newStreamProtocolV4(options)
+
+	creator := newWSStreamCreator(conn)
+	go creator.run()
+	return streamer.stream(creator)
+}
+
+type wsStreamCreator struct {
+	conn      *gwebsocket.Conn
+	connMu    sync.Mutex
+	streams   map[byte]*stream
+	streamsMu sync.Mutex
+}
+
+func newWSStreamCreator(conn *gwebsocket.Conn) *wsStreamCreator {
+	ws := wsStreamCreator{
+		conn:    conn,
+		streams: map[byte]*stream{},
+	}
+
+	go ws.sendPings(PingPeriod) // start heartbeat
+
+	return &ws
+}
+
+func (c *wsStreamCreator) getStream(id byte) *stream {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	return c.streams[id]
+}
+
+func (c *wsStreamCreator) setStream(id byte, s *stream) {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	c.streams[id] = s
+}
+
+// CreateStream uses id from passed headers to create a stream over "c.conn" connection.
+// Returns a Stream structure or nil and an error if one occurred.
+func (c *wsStreamCreator) CreateStream(headers http.Header) (httpstream.Stream, error) {
+	streamType := headers.Get(v1.StreamType)
+	id, ok := streamType2streamID[streamType]
+	if !ok {
+		return nil, fmt.Errorf("unknown stream type: %s", streamType)
+	}
+	if s := c.getStream(id); s != nil {
+		return nil, fmt.Errorf("duplicate stream for type %s", streamType)
+	}
+	reader, writer := io.Pipe()
+	s := &stream{
+		headers:   headers,
+		readPipe:  reader,
+		writePipe: writer,
+		conn:      c.conn,
+		connMu:    &c.connMu,
+		id:        id,
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+// sendPings starts heartbeat sending a "ping" from the endpoint every "period".
+func (c *wsStreamCreator) sendPings(period time.Duration) {
+	c.connMu.Lock()
+	closedCh := make(chan struct{})
+	closeHandler := c.conn.CloseHandler()
+	c.conn.SetCloseHandler(func(code int, text string) error {
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+func (c *wsStreamCreator) run() {
+	readBuffer := make([]byte, remotecommand.WebSocketBufferSize)
+	// Set up handler for ping/pong heartbeat.
+	if err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline)); err != nil {
+		klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+	}
+	c.conn.SetPongHandler(func(string) error {
+		err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline))
+		if err != nil {
+			klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+		}
+		return nil
```

Comment:
I don't fully understand this logic with deadlines, so the logic should be if we sent N pings and we don't receive any pong we should tear down the connections ... the deadlines help to avoid deadlocks, since the read and write operations are blocking this will unblocked them

Reply:
There is a discussion with respect to applying heartbeats in gorilla/websocket https://github.com/gorilla/websocket/issues/304 and the implementation of heartbeats in this PR aligns with the example https://github.com/gorilla/websocket/blob/main/examples/chat/client.go

Reply:
yeah, but both examples close the connection on the pong handlers https://github.com/gorilla/websocket/blob/76ecc29eff79f0cedf70c530605e486fc32131d1/examples/command/main.go#L43

What I fail to see is how this logic close the connection if pings and/or pong are not received, ie, how this detect and idle connection and tears it down

Reply:
I believe this is a non-issue. We did not get the opportunity to resolve this before Antonio went on vacation. The way the heartbeat works for the gorilla websockets is the read deadline is set at the beginning of the read loop and reset upon receiving each pong. If the pong is not received before the read deadline, the connection `NextReader` will return an error. This mechanism is discussed in this thread: https://stackoverflow.com/questions/37696527/go-gorilla-websockets-on-ping-pong-fail-user-disconnct-call-function.